Let toolsets be built dynamically based on run context

strawgate opened the pull request:
This PR adds a decorator that can be used to dynamically add toolsets to an Agent run similar to how the system prompt and instructions can be added to via decorated functions.

- Closes https://github.com/pydantic/pydantic-ai/issues/2331

DouweM commented:
@strawgate Thanks! My main concern is that the function taking `ctx` suggests that a new toolset is going to be built when the context changes (on each run step) or at least when a new run is performed (and the `ctx.deps` would likely change), so that you can safely use run/user-specific details from `ctx.deps` like in your filesystem root directory example, without accidentally using a toolset built for another user/run. 

But in reality it is tied to toolset enters/exits and there's no such guarantee... Those _could_ align with agent runs, but wouldn't if you have your own `async with agent:` wrapping multiple agent runs (which you may want to do to keep MCP servers running during an entire request that could include multiple agent runs) because of the `entered_count` stuff we do in `Agent.__aenter__` and `CombinedToolset.__aenter__`. Maybe we should just rip that out, so that we know for sure we enter/exit each toolset whenever a new agent run starts? Maybe we should also pass only `deps` instead of `RunContext` to make it clear you're not actually getting run-step specific context?

I want to make sure we don't accidentally introduce a footgun where the behavior will change completely when you add an `async with agent:` around a request or the entire FastAPI lifecycle or something, thinking you're just optimizing e.g. MCP server start/shutdowns.

Maybe we need to add a new hook to toolsets that's only called when a new run/step is actually started, so we don't rely on the enters/exits which could happen at any level?

This is all a bit complicated and my analysis here may very well be wrong. Either way I think we'll need tests with different patterns to make sure we get the desired behavior in each case.

We should probably chat about this directly, would you mind joining [our public Slack](https://logfire.pydantic.dev/docs/join-slack/) so we can huddle?

strawgate commented:
@DouweM i added the toolset decorator and added toolsetfunc to agent init, the test appears to be passing, I'll play around with this a little bit more though

DouweM commented on pydantic_ai_slim/pydantic_ai/agent.py:
```diff
@@ -421,7 +422,9 @@ def __init__(
         self._function_toolset = FunctionToolset(tools, max_retries=retries)
-        self._user_toolsets = toolsets or ()
+        self._user_toolsets = [
+            toolset if isinstance(toolset, AbstractToolset) else DynamicToolset(toolset) for toolset in toolsets or []
```
Whoops, this `DynamicToolset` would be shared across multiple uses of the agent (possibly with different deps for different users), so this won't work right. We're better off storing the functions as you had it, and then building the `DynamicToolset` inside `iter` (as you had it). And making `DynamicToolset` private as we don't actually want people to create it themselves and run into this issue.

Or we'd need to fix `DynamicToolset` to keep a separate toolset per deps, but then deps would need to be hashable, which is not something we want to require.

DouweM replied:
@strawgate I think it'll work if we only build the `DynamicToolset` inside `iter` and pass it to `ToolManager` there (via the `_get_toolset` call as you were doing before), so every run will get its own toolset.

strawgate replied:
I looked at this a bit and I'm wondering if the simplest solution is to just add a copy method to the dynamic toolset and just make a copy of it in iter / _get_toolset

will make a commit for this to look at

strawgate commented:
looks like unrelated test is failing

DouweM commented:
@strawgate I made some tweaks to the docs, having this replace the `WrapperToolset` workaround for dynamic toolsets that encouraged some bad patterns. If this all looks good to you I think we can merge soon!

strawgate commented:
Looks great

HamzaFarhan commented:
I feel like after this it makes even more sense for `output_type` to also accept toolsets
Right now if we have 10 output_types, and 5 toolsets with let's say 5-6 tools per toolset, we can add a toolset dynamically. In other words, add 5-6 tools with a single check. But we would need to add the output_types using if else in a single prepare function.
If they were toolsets, the same trigger could be used to load 5-6 tools and 2-3 corresponding output_types.

DouweM commented:
@HamzaFarhan For proper type checking of the `agent.result.output`, the possible output types need to be statically known at agent definition time, which wouldn't work with toolsets. Type safety is pretty core to Pydantic (and AI's) philosophy, so that makes this a non-starter unfortunately.
